src/CMakeLists.txt

```diff
@@ -113,20 +208,12 @@ target_link_libraries(
     $<$<BOOL:${CP2K_USE_COSMA}>:cosma::cosma_prefixed_pxgemm>
     $<$<BOOL:${CP2K_USE_COSMA}>:cosma::cosma>
     DBCSR::dbcsr
-    cp2k_grid
-    cp2k_dbm
-    $<$<BOOL:${CP2K_USE_CUDA}>:CUDA::cufft>
-    $<$<BOOL:${CP2K_USE_CUDA}>:CUDA::cufftw>
-    $<$<BOOL:${CP2K_USE_CUDA}>:CUDA::cublas>
-    $<$<BOOL:${CP2K_USE_CUDA}>:CUDA::cudart>
-    $<$<BOOL:${CP2K_USE_HIP}>:hip::hipfft>
-    $<$<BOOL:${CP2K_USE_HIP}>:roc::hipblas>
-    $<$<BOOL:${CP2K_USE_HIP}>:hip::host>
     $<$<BOOL:${CP2K_USE_MPI}>:CP2K::SCALAPACK::scalapack>
-    $<$<BOOL:${CP2K_USE_LIBXSMM}>:CP2K::LibXSMM::libxsmmf>
-    $<$<BOOL:${CP2K_USE_LIBXSMM}>:CP2K::LibXSMM::libxsmmext>
     CP2K::LAPACK::lapack
-    CP2K::BLAS::blas
+    cp2k_grid
+    cp2k_dbm
+    $<$<BOOL:${CP2K_USE_CUDA}>:cp2k_cuda_libs>
```

Comment:
Do we actually need [generator expressions](https://cmake.org/cmake/help/v3.0/manual/cmake-generator-expressions.7.html)? It seems they could all be replaced by much more readable if-constructs like this:


```
if(CP2K_USE_CUDA)
  target_link_libraries(cp2k_link_libs INTERFACE cp2k_cuda_libs)
endif()
```

Reply:
Can you maybe elaborate a bit on why? I assume we both agree that they are taking quite a toll on the readability?

Reply:
This case is clear. Combination of them might be a bit more complicated to understand at first but it is compact and actually a very useful feature of the language.
Basically, I can not use language features I see fit for the job. I get that but these generator will stay as having a sequence of if() endif() is equally tedious to write and read. The generators expression were probably partially introduced to include such things in a unique function call.

I have multiple of these expression in the cmakelists.txt and I really do not see the point of translating them into if statements unless the expression is really complicated `$<$<AND:$<>,$<>>:>` type for instance.

Reply:
I'm sorry, but I did not find `$<$<BOOL:${CP2K_USE_CUDA}>:cp2k_cuda_libs>` clear. The syntax does not resemble any other language I'm aware of. 

Besides the syntax the larger problem with this compact style is that it's hard to reason about.
The code would be much easier to understand if it were organized like this:
```
if(FOO)
  # All the effects of FOO.
endif()

if(BAR)
  # All the effects of BAR.
endif()

if(FOO AND BAR)
 # All the effects of FOO and BAR interacting.
endif()
```

Reply:
> I'm sorry, but I did not find `$<$<BOOL:${CP2K_USE_CUDA}>:cp2k_cuda_libs>` clear. The syntax does not resemble any other language I'm aware of.
> 
> Besides the syntax the larger problem with this compact style is that it's hard to reason about. The code would be much easier to understand if it were organized like this:
> 
> ```
> if(FOO)
>   # All the effects of FOO.
> endif()
> 
> if(BAR)
>   # All the effects of BAR.
> endif()
> 
> if(FOO AND BAR)
>  # All the effects of FOO and BAR interacting.
> endif()
> ```


$<$<BOOL:${CP2K_USE_CUDA}>:cp2k_cuda_libs> is clear. It just mean check the value of CP2K_USE_CUDA (BOOL is to indicate the type) and if true add the target otherwise return an empty string. Really I have seen far more complicated syntax than this. the reverse polish notation been one.

The cmake community uses generator expressions when they can. Generators expressions have many advantages compared to the conventional if - then -else statement. it is possible to create a list of them and just pass the list to any functions that supports this functionality. The target functions are a typical use case.

Reply:
For the generator expressions: When in Rome, Do as the Romans Do. They are everywhere in modern CMake and some things (like object lib dependencies and the like) are likely not possible anymore without them. For better readability it makes sense to [distribute them over multiple lines](https://github.com/cp2k/dbcsr/blob/0a78582d8aed3f53915c101773e5a535e16770d8/src/CMakeLists.txt#L231-L254) and use indentation accordingly.

And when looking at Bash variable substitution or Ansible (Jinja2) filter expressions you will see that besides the different bracketing style they are not that uncommon.

Reply:
yes indentation is important (from a visual point of view). I will check again that I did not miss one or two.

Reply:
> some things (like object lib dependencies and the like) are likely not possible anymore without them. 

Ok, if they are actually unavoidable then so be it.